pandas/plotting/_matplotlib/core.py

```diff
@@ -240,18 +246,30 @@ def _iter_data(self, data=None, keep_index=False, fillna=None):
         if fillna is not None:
             data = data.fillna(fillna)
 
-        for col, values in data.items():
-            if keep_index is True:
-                yield col, values
-            else:
-                yield col, values.values
+        if self.by is None:
```

Comment:
A lot of code in this and the else block looks duplicated - is there a way to consolidate?

Reply:
emm, this could leave here i think, because the only duplicated part is to yield values. And there is a reason for those two to use different structure, (the first case uses `data.items` while the second iterates columns), this is initially brought in #10962 to solve a corner case, however, we do not need this for `by` case since this won't happen (duplicated column names), so I think it is fine to just leave as is. Opinions are welcome!

Reply:
IMO can remove some of the duplication - could be written as
```python
for col, values in data.items():
     if not keep_index:
          values = value.values
     yield col, values
```
Will get rid of 4 LOC